cilium/cmd/kernel_check.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"text/tabwriter"
+
+	"github.com/cilium/cilium/pkg/command"
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/spf13/cobra"
+)
+
+var kernelCheckCmd = &cobra.Command{
+	Use:   "kernel-check",
+	Short: "Checks whether machine is supporting required kernel capabilities",
+	Run: func(cmd *cobra.Command, args []string) {
+		kernelCheck()
+	},
+}
+
+type KernelParamSupport struct {
+	feature   probes.KernelParam
+	supported bool
+	required  bool
+}
+
+func kernelCheck() {
+	probeManager := probes.NewProbeManager()
+	w := tabwriter.NewWriter(os.Stdout, 30, 8, 0, ' ', 0)
+	if err := probeManager.ExitKernelCheck(); err != nil {
+		Fatalf("%s", err)
+	} else {
+		fmt.Fprintln(w, "FEATURE\tSUPPORTED\tREQUIRED")
+		requiredParams := probeManager.GetRequiredConfig()
+		listParams := []KernelParamSupport{}
+		for f, s := range requiredParams {
+			listParams = append(listParams, KernelParamSupport{
+				feature:   f,
+				supported: s,
+				required:  true,
+			})
+		}
+		optionalParams := probeManager.GetOptionalConfig()
+		for f, s := range optionalParams {
+			listParams = append(listParams, KernelParamSupport{
+				feature:   f,
+				supported: s,
+				required:  false,
+			})
+		}
+		for _, p := range listParams {
+			fmt.Fprintf(w, "%s\t%t\t%t\n", p.feature, p.supported, p.required)
+		}
+		if command.OutputJSON() {
+			json.Marshal(listParams)
+			if err := command.PrintOutput(listParams); err != nil {
+				Fatalf("Unable to generate JSON output: %s", err)
+			}
+			return
+		}
+		w.Flush()
+	}
+}
```

Comment:
With the fix on struct `KernelParamSupport` fields, and with a bit of reorganisation to keep all non-JSON stuff together (and to avoid filling the writer if we want JSON and won't flush it ever), it could look like the following. Don't hesitate to return early to avoid nested block (You do it for JSON printing; also, no need for the `else` since a fail on `ExitKernelCheck()` exits with `Fatalf()` anyway).

```suggestion
type KernelParamSupport struct {
	Feature   probes.KernelParam
	Supported bool
	Required  bool
}

func kernelCheck() {
	probeManager := probes.NewProbeManager()
	if err := probeManager.ExitKernelCheck(); err != nil {
		Fatalf("%s", err)
	}

	listParams := []KernelParamSupport{}
	requiredParams := probeManager.GetRequiredConfig()
	for f, s := range requiredParams {
		listParams = append(listParams, KernelParamSupport{
			Feature:   f,
			Supported: s,
			Required:  true,
		})
	}
	optionalParams := probeManager.GetOptionalConfig()
	for f, s := range optionalParams {
		listParams = append(listParams, KernelParamSupport{
			Feature:   f,
			Supported: s,
			Required:  false,
		})
	}

	if command.OutputJSON() {
		if err := command.PrintOutput(listParams); err != nil {
			Fatalf("Unable to generate JSON output: %s", err)
			os.Exit(1)
		}
		return
	}

	w := tabwriter.NewWriter(os.Stdout, 30, 8, 0, ' ', 0)
	fmt.Fprintln(w, "FEATURE\tSUPPORTED\tREQUIRED")
	for _, p := range listParams {
		fmt.Fprintf(w, "%s\t%t\t%t\n", p.Feature, p.Supported, p.Required)
	}
	w.Flush()
}
```